internal/exec/exec.go

```diff
@@ -0,0 +1,167 @@
+package exec
+
+import (
+	"fmt"
+	"os/exec"
+
+	"github.com/apex/log"
+	"github.com/goreleaser/goreleaser/internal/artifact"
+	"github.com/goreleaser/goreleaser/internal/logext"
+	"github.com/goreleaser/goreleaser/internal/pipe"
+	"github.com/goreleaser/goreleaser/internal/semerrgroup"
+	"github.com/goreleaser/goreleaser/internal/tmpl"
+	"github.com/goreleaser/goreleaser/pkg/config"
+	"github.com/goreleaser/goreleaser/pkg/context"
+	"github.com/mattn/go-shellwords"
+)
+
+func Execute(ctx *context.Context, publishers []config.Publisher) error {
+	if ctx.SkipPublish {
+		return pipe.ErrSkipPublishEnabled
+	}
+
+	for _, p := range publishers {
+		log.WithField("name", p.Name).Debug("executing custom publisher")
+		err := executePublisher(ctx, p)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func executePublisher(ctx *context.Context, publisher config.Publisher) error {
+	log.Debugf("filtering %d artifacts", len(ctx.Artifacts.List()))
+	artifacts := filterArtifacts(ctx.Artifacts, publisher)
+	log.Debugf("will execute custom publisher with %d artifacts", len(artifacts))
+
+	var g = semerrgroup.New(ctx.Parallelism)
+	for _, artifact := range artifacts {
+		artifact := artifact
+		g.Go(func() error {
+			c, err := resolveCommand(ctx, publisher, artifact)
+			if err != nil {
+				return err
+			}
+
+			return executeCommand(c)
+		})
+	}
+
+	return g.Wait()
+}
+
+func executeCommand(c *command) error {
+	log.WithField("args", c.Args).
+		WithField("env", c.Env).
+		Debug("executing command")
+
+	// nolint: gosec
+	var cmd = exec.CommandContext(c.Ctx, c.Args[0], c.Args[1:]...)
+	cmd.Env = c.Env
+	if c.Dir != "" {
+		cmd.Dir = c.Dir
+	}
+
+	entry := log.WithField("cmd", c.Args[0])
+	cmd.Stderr = logext.NewErrWriter(entry)
+	cmd.Stdout = logext.NewWriter(entry)
+
+	log.WithField("cmd", cmd.Args).Info("publishing")
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("publishing: %s failed: %w",
+			c.Args[0], err)
+	}
+
+	log.Debugf("command %s finished successfully", c.Args[0])
+	return nil
+}
+
+func filterArtifacts(artifacts artifact.Artifacts, publisher config.Publisher) []*artifact.Artifact {
+	var filter artifact.Filter
+	filters := append([]artifact.Filter{},
+		artifact.ByType(artifact.UploadableArchive),
+		artifact.ByType(artifact.UploadableFile),
+		artifact.ByType(artifact.LinuxPackage),
+		artifact.ByType(artifact.UploadableBinary))
```

Comment:
could be simplified to:
```suggestion
	filters := []artifact.Filter{
		artifact.ByType(artifact.UploadableArchive),
		artifact.ByType(artifact.UploadableFile),
		artifact.ByType(artifact.LinuxPackage),
		artifact.ByType(artifact.UploadableBinary),
	}
```

Reply:
Indeed - I don't know why did I do it in such complicated way! 😄